File: desktop/news/calendar.py
"""Economic calendar. Mirrors src/lib/news/calendar.ts.

With FMP_API_KEY set, uses FMP's live calendar (real forecast/previous/actual,
including FOMC/Fed events). Without a key, generates the well-known *recurring*
U.S. releases (CPI, PPI, PCE, NFP, Retail Sales, Consumer Confidence, PMI) from
their standard publication rules and marks them "estimated" — one-off dates
(FOMC meetings, Fed speakers, Treasury auctions) are deliberately not fabricated
since they aren't derivable from a fixed rule; those only appear via a live feed.
"""
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from typing import Callable, Optional

from .sources import fetch_fmp_economic_calendar
from .types import EconomicEvent, Impact


def _nth_weekday(year: int, month: int, weekday: int, n: int) -> datetime:
    """weekday: Monday=0 ... Sunday=6 (Python convention)."""
    d = datetime(year, month, 1, tzinfo=timezone.utc)
    count = 0
    while True:
        if d.weekday() == weekday:
            count += 1
            if count == n:
                return d
        d += timedelta(days=1)


def _last_weekday(year: int, month: int, weekday: int) -> datetime:
    if month == 12:
        d = datetime(year, 12, 31, tzinfo=timezone.utc)
    else:
        d = datetime(year, month + 1, 1, tzinfo=timezone.utc) - timedelta(days=1)
    while d.weekday() != weekday:
        d -= timedelta(days=1)
    return d


def _nth_business_day(year: int, month: int, n: int) -> datetime:
    d = datetime(year, month, 1, tzinfo=timezone.utc)
    count = 0
    while True:
        if d.weekday() < 5:  # Mon-Fri
            count += 1
            if count == n:
                return d
        d += timedelta(days=1)


@dataclass
class _RecurringDef:
    event: str
    importance: int
    impact: Impact
    at: Callable[[int, int], datetime]


# weekday indices: JS Sunday=0..Saturday=6; Python Monday=0..Sunday=6.
# Friday(JS 5) == Friday(Py 4); Tuesday(JS 2) == Tuesday(Py 1).
RECURRING: list[_RecurringDef] = [
    _RecurringDef("Nonfarm Payrolls / Unemployment Rate", 3, "High", lambda y, m: _nth_weekday(y, m, 4, 1)),
    _RecurringDef("CPI (Consumer Price Index)", 3, "High", lambda y, m: _nth_business_day(y, m, 10)),
    _RecurringDef("PPI (Producer Price Index)", 2, "Medium", lambda y, m: _nth_business_day(y, m, 11)),
    _RecurringDef("PCE Price Index", 3, "High", lambda y, m: _last_weekday(y, m, 4)),
    _RecurringDef("Retail Sales", 2, "Medium", lambda y, m: _nth_business_day(y, m, 14)),
    _RecurringDef("ISM Manufacturing PMI", 2, "Medium", lambda y, m: _nth_business_day(y, m, 1)),
    _RecurringDef("ISM Services PMI", 2, "Medium", lambda y, m: _nth_business_day(y, m, 3)),
    _RecurringDef("Consumer Confidence (Conference Board)", 2, "Medium", lambda y, m: _last_weekday(y, m, 1)),
]


def _estimated_events(range_start: datetime, range_end: datetime) -> list[EconomicEvent]:
    months: set[tuple[int, int]] = set()
    d = range_start
    while d <= range_end:
        months.add((d.year, d.month))
        d += timedelta(days=1)

    events: list[EconomicEvent] = []
    for year, month in months:
        for definition in RECURRING:
            date = definition.at(year, month)
            if range_start <= date <= range_end:
                event_time = date.replace(hour=12, minute=30)
                events.append(
                    EconomicEvent(
                        id=f"est-{definition.event}-{date:%Y-%m-%d}",
                        time=event_time.isoformat(),
                        country="US",
                        event=definition.event,
                        importance=definition.importance,
                        impact=definition.impact,
                        estimated=True,
                    )
                )
    return sorted(events, key=lambda e: e.time)


def _fmp_impact(raw: Optional[str]) -> Impact:
    v = (raw or "").lower()
    if v == "high":
        return "High"
    if v == "low":
        return "Low"
    return "Medium"


def _fmp_importance(impact: Impact) -> int:
    return {"High": 3, "Medium": 2, "Low": 1}[impact]


@dataclass
class CalendarResult:
    events: list[EconomicEvent]
    live: bool
    error: Optional[str] = None


def load_economic_calendar(range_start: datetime, range_end: datetime) -> CalendarResult:
    """Fetches the calendar for [range_start, range_end] (inclusive), live where possible."""
    live = fetch_fmp_economic_calendar(f"{range_start:%Y-%m-%d}", f"{range_end:%Y-%m-%d}")

    if not live["configured"]:
        return CalendarResult(events=_estimated_events(range_start, range_end), live=False)
    if not live["ok"]:
        return CalendarResult(events=_estimated_events(range_start, range_end), live=False, error=live.get("error"))

    events: list[EconomicEvent] = []
    for idx, i in enumerate(live["items"]):
        if i.get("country") != "US":
            continue
        impact = _fmp_impact(i.get("impact"))
        events.append(
            EconomicEvent(
                id=f"fmp-{idx}-{i['date']}",
                time=datetime.fromisoformat(i["date"]).isoformat(),
                country=i["country"],
                event=i["event"],
                forecast=str(i["estimate"]) if i.get("estimate") is not None else None,
                previous=str(i["previous"]) if i.get("previous") is not None else None,
                actual=str(i["actual"]) if i.get("actual") is not None else None,
                importance=_fmp_importance(impact),
                impact=impact,
                estimated=False,
            )
        )
    return CalendarResult(events=sorted(events, key=lambda e: e.time), live=True)

File: src/lib/news/calendar.ts
/**
 * Economic calendar. When VITE_FMP_API_KEY is set we use FMP's live calendar (real
 * forecast/previous/actual values, including FOMC/Fed events). Without a key we still
 * want a useful calendar, so we generate the well-known *recurring* U.S. releases
 * (CPI, PPI, PCE, NFP, Retail Sales, Consumer Confidence, PMI) from their standard
 * publication rules and mark them "estimated" — we deliberately do NOT fabricate
 * one-off dates (FOMC meeting dates, Fed speaker calendars, Treasury auction
 * schedules) since those aren't derivable from a fixed rule and guessing them would
 * be misleading; those three surface only from the live FMP/Finnhub feed.
 */
import { fetchFmpEconomicCalendar } from './sources';
import type { EconomicEvent, Impact } from './types';

function nthWeekday(year: number, month: number, weekday: number, n: number): Date {
  const d = new Date(Date.UTC(year, month, 1));
  let count = 0;
  while (true) {
    if (d.getUTCDay() === weekday) {
      count += 1;
      if (count === n) return d;
    }
    d.setUTCDate(d.getUTCDate() + 1);
  }
}

function lastWeekday(year: number, month: number, weekday: number): Date {
  const d = new Date(Date.UTC(year, month + 1, 0));
  while (d.getUTCDay() !== weekday) d.setUTCDate(d.getUTCDate() - 1);
  return d;
}

function nthBusinessDay(year: number, month: number, n: number): Date {
  const d = new Date(Date.UTC(year, month, 1));
  let count = 0;
  while (true) {
    const day = d.getUTCDay();
    if (day !== 0 && day !== 6) {
      count += 1;
      if (count === n) return d;
    }
    d.setUTCDate(d.getUTCDate() + 1);
  }
}

interface RecurringDef {
  event: string;
  importance: 1 | 2 | 3;
  impact: Impact;
  at: (year: number, month: number) => Date;
}

const RECURRING: RecurringDef[] = [
  { event: 'Nonfarm Payrolls / Unemployment Rate', importance: 3, impact: 'High', at: (y, m) => nthWeekday(y, m, 5, 1) },
  { event: 'CPI (Consumer Price Index)', importance: 3, impact: 'High', at: (y, m) => nthBusinessDay(y, m, 10) },
  { event: 'PPI (Producer Price Index)', importance: 2, impact: 'Medium', at: (y, m) => nthBusinessDay(y, m, 11) },
  { event: 'PCE Price Index', importance: 3, impact: 'High', at: (y, m) => lastWeekday(y, m, 5) },
  { event: 'Retail Sales', importance: 2, impact: 'Medium', at: (y, m) => nthBusinessDay(y, m, 14) },
  { event: 'ISM Manufacturing PMI', importance: 2, impact: 'Medium', at: (y, m) => nthBusinessDay(y, m, 1) },
  { event: 'ISM Services PMI', importance: 2, impact: 'Medium', at: (y, m) => nthBusinessDay(y, m, 3) },
  { event: 'Consumer Confidence (Conference Board)', importance: 2, impact: 'Medium', at: (y, m) => lastWeekday(y, m, 2) },
];

function estimatedEvents(rangeStart: Date, rangeEnd: Date): EconomicEvent[] {
  const events: EconomicEvent[] = [];
  const months = new Set<string>();
  for (const d = new Date(rangeStart); d <= rangeEnd; d.setUTCDate(d.getUTCDate() + 1)) {
    months.add(`${d.getUTCFullYear()}-${d.getUTCMonth()}`);
  }
  for (const key of months) {
    const [y, m] = key.split('-').map(Number);
    for (const def of RECURRING) {
      const date = def.at(y, m);
      if (date >= rangeStart && date <= rangeEnd) {
        events.push({
          id: `est-${def.event}-${date.toISOString().slice(0, 10)}`,
          time: new Date(Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate(), 12, 30)).toISOString(),
          country: 'US',
          event: def.event,
          importance: def.importance,
          impact: def.impact,
          estimated: true,
        });
      }
    }
  }
  return events.sort((a, b) => +new Date(a.time) - +new Date(b.time));
}

function fmpImpact(raw?: string): Impact {
  const v = (raw ?? '').toLowerCase();
  if (v === 'high') return 'High';
  if (v === 'low') return 'Low';
  return 'Medium';
}

function fmpImportance(impact: Impact): 1 | 2 | 3 {
  return impact === 'High' ? 3 : impact === 'Medium' ? 2 : 1;
}

export interface CalendarResult {
  events: EconomicEvent[];
  live: boolean;
  error?: string;
}

/** Fetches the calendar for [rangeStart, rangeEnd] (inclusive), live where possible. */
export async function loadEconomicCalendar(rangeStart: Date, rangeEnd: Date): Promise<CalendarResult> {
  const fmt = (d: Date) => d.toISOString().slice(0, 10);
  const live = await fetchFmpEconomicCalendar(fmt(rangeStart), fmt(rangeEnd));

  if (!live.configured) {
    return { events: estimatedEvents(rangeStart, rangeEnd), live: false };
  }
  if (!live.ok) {
    return { events: estimatedEvents(rangeStart, rangeEnd), live: false, error: live.error };
  }
  const events: EconomicEvent[] = live.items
    .filter((i) => i.country === 'US')
    .map((i, idx) => {
      const impact = fmpImpact(i.impact);
      return {
        id: `fmp-${idx}-${i.date}`,
        time: new Date(i.date).toISOString(),
        country: i.country,
        event: i.event,
        forecast: i.estimate !== undefined ? String(i.estimate) : undefined,
        previous: i.previous !== undefined ? String(i.previous) : undefined,
        actual: i.actual !== undefined ? String(i.actual) : undefined,
        importance: fmpImportance(impact),
        impact,
        estimated: false,
      };
    });
  return { events: events.sort((a, b) => +new Date(a.time) - +new Date(b.time)), live: true };
}
